### ISSUE TITLE
wxCaret doesn't work in wxGTK3 (was: wayland breaks wxClientDC)

### ISSUE THREAD

**Issue by wxtrac** [wxWidgets/wxWidgets#17820]
*Issue migrated from trac ticket # 17820*

**component:** wxGTK | **priority:** normal | **keywords:** Wayland

#### 2017-03-12 13:43:00: e9925248 created the issue

___
Compile the caret sample from the wxWidgets sources on Fedora 25:
a) Using X11: Everything works
b) Using Wayland and GDK_BACKEND=x11 (+Xwayland): Typing the next character draws a black block - moving the blinking caret seems to leave the previous position black.
c) Using Wayland and GDK_BACKEND=wayland: The cursor does not even move - only a (externally forced) window repaint fixes the window content.

Reason for c: wxClientDC, wxWindowDC and wxScreenDC are based on gdk_cairo_create.

Wayland wants to switch to a "only draw during paint events" - so there has probably no interest in forwarding content drawn on the cairo context created by gdk_cairo_create.

This function is already deprecated
https://developer.gnome.org/gdk3/stable/gdk3-Cairo-Interaction.html#gdk-window-create-similar-surface
and will be removed in future versions (3.89.X):
https://git.gnome.org/browse/gtk+/commit/?id=d48adf9cee7e340acd7f8b9a5f9716695352b848

Reference: https://bugzilla.redhat.com/show_bug.cgi?id=1431233

**Comment by vadz** [comment]
AFAICS #2522 provides the solution for this issue, so we need to merge it before 3.1.6 to fix this before 3.2.0.

**Comment by vadz** [comment]
@paulcor Do you think it's worth making an effort (and taking time) to merge #2522 before 3.1.6 or not? Because considering the amount of changes there, I won't risk merging it after 3.1.6, i.e. if we don't do it now, it won't be in 3.2.0.

**Comment by paulcor** [comment]
There is so much in #2522 that it is really hard to review. I'd like to make changes more incrementally, with an initial goal of just getting `wxOverlay` to work with Wayland. I can dedicate some time, so my intention is to work towards that goal over the next few days.

**Comment by vadz** [comment]
FWIW the symptoms are slightly different, but the caret sample still doesn't work with Wayland, as it still uses `wxClientDC` and not `wxOverlay`. I thought some of @AliKet's PRs already updated it to use the latter, but I can't find anything right now (it's not part of #2522), so I'll probably try doing it myself.

But if we're just going to update the sample, this is probably not 3.1.6-critical any more...

**Comment by paulcor** [comment]
> FWIW the symptoms are slightly different, but the caret sample still doesn't work with Wayland

It seems to work fine for me with Weston, and I have also tested with Gnome. The previous "symptom" was that `wxCaret` didn't work at all, so what symptom is "slightly different" than that?

> as it still uses `wxClientDC` and not `wxOverlay`

For Wayland, it uses `wxOverlay`.

**Comment by vadz** [comment]
I don't see any typed text at all when using Wayland, even after refreshing/resizing the window. This is with bog standard GNOME installation.

**Comment by paulcor** [comment]
Ah, my mistake, I only tested that the **caret** worked, not typed text. But that is trivial to fix:
```
diff --git a/samples/caret/caret.cpp b/samples/caret/caret.cpp
index 815cab4527..1e8581a3a9 100644
--- a/samples/caret/caret.cpp
+++ b/samples/caret/caret.cpp
@@ -473,12 +473,12 @@ void MyCanvas::OnChar( wxKeyEvent &event )
                 wxChar ch = (wxChar)event.GetKeyCode();
                 CharAt(m_xCaret, m_yCaret) = ch;
 
-                wxCaretSuspend cs(this);
-                wxClientDC dc(this);
-                dc.SetFont(m_font);
-                dc.SetBackgroundMode(wxBRUSHSTYLE_SOLID); // overwrite old value
-                dc.DrawText(ch, m_xMargin + m_xCaret * m_widthChar,
-                                m_yMargin + m_yCaret * m_heightChar );
+                wxRect rect(
+                    m_xMargin + m_xCaret * m_widthChar,
+                    m_yMargin + m_yCaret * m_heightChar,
+                    m_widthChar,
+                    m_heightChar);
+                Refresh(false, &rect);
 
                 NextChar();
             }
```

**Comment by vadz** [comment]
Thanks, I've committed this fix and it does work fine under Wayland now.

There are still cosmetic problems with X11 backend, ironically: there seems to be off by 1 error somewhere as the upper and left sides of the caret rectangle remain shown after the caret is moved, so I keep this one open, but reset the milestone as it doesn't seem urgent to fix this.

**Comment by paulcor** [comment]
> there seems to be off by 1 error somewhere as the upper and left sides of the caret rectangle remain shown

I think that is fixed by 03bf61be3c

**Comment by vadz** [comment]
Closing as this seems to have been fixed and I don't see any problems in the caret sample when testing it under neither X nor Wayland.
